File: src/core/integrations/ollama.ts
import { logger, PerformanceLogger } from '../../utils/logger';
import { OllamaError, ErrorCode } from '../../utils/errors';

/**
 * Ollama API Client
 * Provides interface to local Ollama instance for LLM inference
 */

// ============================================================================
// Types
// ============================================================================

export interface OllamaConfig {
  endpoint: string; // e.g., "http://localhost:11434"
  timeout?: number; // milliseconds, default 30000
}

export interface OllamaModel {
  name: string;
  modified_at: string;
  size: number;
  digest: string;
  details?: {
    format?: string;
    family?: string;
    families?: string[];
    parameter_size?: string;
    quantization_level?: string;
  };
}

export interface OllamaGenerateRequest {
  model: string;
  prompt: string;
  system?: string;
  template?: string;
  context?: number[];
  stream?: boolean;
  raw?: boolean;
  format?: 'json';
  options?: {
    temperature?: number;
    top_p?: number;
    top_k?: number;
    num_predict?: number;
    stop?: string[];
  };
}

export interface OllamaGenerateResponse {
  model: string;
  created_at: string;
  response: string;
  done: boolean;
  context?: number[];
  total_duration?: number;
  load_duration?: number;
  prompt_eval_count?: number;
  prompt_eval_duration?: number;
  eval_count?: number;
  eval_duration?: number;
}

export interface OllamaChatMessage {
  role: 'system' | 'user' | 'assistant';
  content: string;
  images?: string[]; // base64 encoded images
}

export interface OllamaChatRequest {
  model: string;
  messages: OllamaChatMessage[];
  stream?: boolean;
  format?: 'json';
  options?: {
    temperature?: number;
    top_p?: number;
    top_k?: number;
    num_predict?: number;
  };
}

export interface OllamaChatResponse {
  model: string;
  created_at: string;
  message: OllamaChatMessage;
  done: boolean;
  total_duration?: number;
  load_duration?: number;
  prompt_eval_count?: number;
  prompt_eval_duration?: number;
  eval_count?: number;
  eval_duration?: number;
}

export interface OllamaHealthResponse {
  status: 'ok' | 'error';
  version?: string;
}

// ============================================================================
// Ollama Client
// ============================================================================

export class OllamaClient {
  private config: Required<OllamaConfig>;

  constructor(config: OllamaConfig) {
    this.config = {
      endpoint: config.endpoint,
      timeout: config.timeout ?? 30000,
    };

    logger.info('Ollama client initialized', {
      endpoint: this.config.endpoint,
    });
  }

  /**
   * Test connection to Ollama instance
   */
  async healthCheck(): Promise<OllamaHealthResponse> {
    const perf = new PerformanceLogger(logger, 'Ollama health check');

    try {
      const response = await this.fetch('/api/version', {
        method: 'GET',
      });

      if (!response.ok) {
        throw new OllamaError(
          ErrorCode.OLLAMA_CONNECTION_FAILED,
          `Health check failed: ${response.status} ${response.statusText}`
        );
      }

      const data = await response.json();
      perf.end({ status: 'ok' });

      return {
        status: 'ok',
        version: data.version,
      };
    } catch (error) {
      perf.endWithError(error as Error);

      if (error instanceof OllamaError) {
        throw error;
      }

      throw new OllamaError(
        ErrorCode.OLLAMA_CONNECTION_FAILED,
        'Failed to connect to Ollama',
        undefined,
        error as Error
      );
    }
  }

  /**
   * List available models
   */
  async listModels(): Promise<OllamaModel[]> {
    const perf = new PerformanceLogger(logger, 'Ollama list models');

    try {
      const response = await this.fetch('/api/tags', {
        method: 'GET',
      });

      if (!response.ok) {
        throw new OllamaError(
          ErrorCode.OLLAMA_CONNECTION_FAILED,
          `Failed to list models: ${response.status}`
        );
      }

      const data = await response.json();
      const models = data.models || [];

      perf.end({ modelCount: models.length });
      return models;
    } catch (error) {
      perf.endWithError(error as Error);

      if (error instanceof OllamaError) {
        throw error;
      }

      throw new OllamaError(
        ErrorCode.OLLAMA_CONNECTION_FAILED,
        'Failed to list models',
        undefined,
        error as Error
      );
    }
  }

  /**
   * Check if a specific model is available
   */
  async hasModel(modelName: string): Promise<boolean> {
    try {
      const models = await this.listModels();
      return models.some((m) => m.name === modelName);
    } catch (error) {
      logger.warn('Failed to check model availability', { modelName }, error as Error);
      return false;
    }
  }

  /**
   * Generate completion (non-streaming)
   */
  async generate(request: OllamaGenerateRequest): Promise<OllamaGenerateResponse> {
    const perf = new PerformanceLogger(logger, 'Ollama generate');

    try {
      const response = await this.fetch('/api/generate', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ ...request, stream: false }),
      });

      if (!response.ok) {
        throw new OllamaError(
          ErrorCode.OLLAMA_INFERENCE_FAILED,
          `Generation failed: ${response.status}`,
          { model: request.model }
        );
      }

      const data = await response.json();
      perf.end({
        model: request.model,
        responseLength: data.response?.length,
        evalCount: data.eval_count,
      });

      return data;
    } catch (error) {
      perf.endWithError(error as Error, { model: request.model });

      if (error instanceof OllamaError) {
        throw error;
      }

      throw new OllamaError(
        ErrorCode.OLLAMA_INFERENCE_FAILED,
        'Failed to generate completion',
        { model: request.model },
        error as Error
      );
    }
  }

  /**
   * Chat completion (non-streaming)
   */
  async chat(request: OllamaChatRequest): Promise<OllamaChatResponse> {
    const perf = new PerformanceLogger(logger, 'Ollama chat');

    try {
      const response = await this.fetch('/api/chat', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ ...request, stream: false }),
      });

      if (!response.ok) {
        throw new OllamaError(
          ErrorCode.OLLAMA_INFERENCE_FAILED,
          `Chat failed: ${response.status}`,
          { model: request.model }
        );
      }

      const data = await response.json();
      perf.end({
        model: request.model,
        messageCount: request.messages.length,
        evalCount: data.eval_count,
      });

      return data;
    } catch (error) {
      perf.endWithError(error as Error, { model: request.model });

      if (error instanceof OllamaError) {
        throw error;
      }

      throw new OllamaError(
        ErrorCode.OLLAMA_INFERENCE_FAILED,
        'Failed to complete chat',
        { model: request.model },
        error as Error
      );
    }
  }

  /**
   * Pull a model from Ollama library
   */
  async pullModel(modelName: string): Promise<void> {
    logger.info('Pulling model', { modelName });

    try {
      const response = await this.fetch('/api/pull', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ name: modelName, stream: false }),
      });

      if (!response.ok) {
        throw new OllamaError(
          ErrorCode.OLLAMA_MODEL_NOT_FOUND,
          `Failed to pull model: ${response.status}`,
          { modelName }
        );
      }

      logger.info('Model pulled successfully', { modelName });
    } catch (error) {
      logger.error('Failed to pull model', { modelName }, error as Error);

      if (error instanceof OllamaError) {
        throw error;
      }

      throw new OllamaError(
        ErrorCode.OLLAMA_MODEL_NOT_FOUND,
        'Failed to pull model',
        { modelName },
        error as Error
      );
    }
  }

  /**
   * Internal fetch wrapper with timeout
   */
  private async fetch(path: string, init?: RequestInit): Promise<Response> {
    const url = `${this.config.endpoint}${path}`;

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);

    try {
      const response = await fetch(url, {
        ...init,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);
      return response;
    } catch (error) {
      clearTimeout(timeoutId);

      if ((error as Error).name === 'AbortError') {
        throw new OllamaError(
          ErrorCode.OLLAMA_TIMEOUT,
          `Request timed out after ${this.config.timeout}ms`,
          { url, timeout: this.config.timeout }
        );
      }

      throw error;
    }
  }

  /**
   * Update client configuration
   */
  updateConfig(config: Partial<OllamaConfig>): void {
    if (config.endpoint) {
      this.config.endpoint = config.endpoint;
    }
    if (config.timeout !== undefined) {
      this.config.timeout = config.timeout;
    }

    logger.info('Ollama client config updated', this.config);
  }

  /**
   * Get current configuration
   */
  getConfig(): OllamaConfig {
    return { ...this.config };
  }
}

// ============================================================================
// Factory Function
// ============================================================================

/**
 * Create a new Ollama client instance
 */
export function createOllamaClient(config: OllamaConfig): OllamaClient {
  return new OllamaClient(config);
}

// Export singleton instance (can be reconfigured)
export let ollamaClient: OllamaClient | null = null;

/**
 * Initialize the global Ollama client
 */
export function initializeOllamaClient(config: OllamaConfig): OllamaClient {
  ollamaClient = new OllamaClient(config);
  return ollamaClient;
}

/**
 * Get the global Ollama client
 */
export function getOllamaClient(): OllamaClient {
  if (!ollamaClient) {
    throw new OllamaError(
      ErrorCode.OLLAMA_CONNECTION_FAILED,
      'Ollama client not initialized. Call initializeOllamaClient() first.'
    );
  }
  return ollamaClient;
}

File: src-tauri/src/integrations/ollama.rs
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::time::Duration;
use thiserror::Error;

/// Ollama client errors
#[derive(Error, Debug, Serialize)]
pub enum OllamaError {
    #[error("Connection failed: {0}")]
    ConnectionFailed(String),

    #[error("Request timeout")]
    Timeout,

    #[error("Model not found: {0}")]
    ModelNotFound(String),

    #[error("Inference failed: {0}")]
    InferenceFailed(String),

    #[error("Invalid response")]
    InvalidResponse,
}

impl From<reqwest::Error> for OllamaError {
    fn from(err: reqwest::Error) -> Self {
        if err.is_timeout() {
            Self::Timeout
        } else if err.is_connect() {
            Self::ConnectionFailed(err.to_string())
        } else {
            Self::InferenceFailed(err.to_string())
        }
    }
}

pub type OllamaResult<T> = Result<T, OllamaError>;

/// Ollama configuration
#[derive(Debug, Clone)]
pub struct OllamaConfig {
    pub endpoint: String,
    pub timeout_secs: u64,
}

impl Default for OllamaConfig {
    fn default() -> Self {
        Self {
            endpoint: "http://localhost:11434".to_string(),
            timeout_secs: 30,
        }
    }
}

/// Ollama model information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OllamaModel {
    pub name: String,
    pub modified_at: String,
    pub size: i64,
    pub digest: String,
}

/// Response from /api/tags
#[derive(Debug, Deserialize)]
struct TagsResponse {
    models: Vec<OllamaModel>,
}

/// Response from /api/version
#[derive(Debug, Deserialize)]
struct VersionResponse {
    version: String,
}

/// Ollama client for Rust-side operations
pub struct OllamaClient {
    config: OllamaConfig,
    client: Client,
}

impl OllamaClient {
    /// Create a new Ollama client
    pub fn new(config: OllamaConfig) -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(config.timeout_secs))
            .build()
            .expect("Failed to build HTTP client");

        log::info!("Ollama client initialized: {}", config.endpoint);

        Self { config, client }
    }

    /// Test connection to Ollama instance
    pub async fn health_check(&self) -> OllamaResult<String> {
        let url = format!("{}/api/version", self.config.endpoint);

        log::debug!("Health check: {}", url);

        let response = self.client.get(&url).send().await?;

        if !response.status().is_success() {
            return Err(OllamaError::ConnectionFailed(format!(
                "HTTP {}",
                response.status()
            )));
        }

        let data: VersionResponse = response.json().await?;

        log::info!("Ollama health check passed: version {}", data.version);
        Ok(data.version)
    }

    /// List available models
    pub async fn list_models(&self) -> OllamaResult<Vec<OllamaModel>> {
        let url = format!("{}/api/tags", self.config.endpoint);

        log::debug!("Listing models: {}", url);

        let response = self.client.get(&url).send().await?;

        if !response.status().is_success() {
            return Err(OllamaError::ConnectionFailed(format!(
                "HTTP {}",
                response.status()
            )));
        }

        let data: TagsResponse = response.json().await?;

        log::info!("Found {} Ollama models", data.models.len());
        Ok(data.models)
    }

    /// Check if a specific model is available
    pub async fn has_model(&self, model_name: &str) -> bool {
        match self.list_models().await {
            Ok(models) => models.iter().any(|m| m.name == model_name),
            Err(e) => {
                log::warn!("Failed to check model availability: {}", e);
                false
            }
        }
    }

    /// Get the client configuration
    pub fn config(&self) -> &OllamaConfig {
        &self.config
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_health_check() {
        // This test requires Ollama to be running
        let config = OllamaConfig::default();
        let client = OllamaClient::new(config);

        // This will fail if Ollama is not running, which is expected
        let result = client.health_check().await;
        println!("Health check result: {:?}", result);
    }
}
